packages/hooks/src/hooks/useContract.ts

```diff
@@ -1,27 +1,38 @@
 import React from 'react';
 import { Web3Context } from '../Provider';
-import { Contract } from 'ethers';
+import { Contract, ContractInterface } from 'ethers';
 
-export function useContract(address: string, abi) {
+/**
+ * @description
+ * Defines the contract instance on `useState` hook
+ */
+export type ContractInstance<T extends Contract> = T | null | undefined;
+
+/**
+ * @description
+ * The return type of the `useContract` hook
+ */
+export type UseContractHook<T extends Contract> = [
+  T | null | undefined,
+  boolean
+];
+
+export function useContract<T extends Contract>(
+  address: string,
+  abi: ContractInterface
+): UseContractHook<T> {
   const context = React.useContext(Web3Context);
-  const [contract, setContract] = React.useState({});
+  const [contract, setContract] = React.useState<ContractInstance<T>>(null);
   const [isReady, setIsReady] = React.useState(false);
   React.useEffect(() => {
     if (context?.connected) {
       const newContract = new Contract(
         address,
         abi,
         context.signer || undefined
-      );
-      const contractInterface = Object.values(
-        newContract.interface.functions
-      ).reduce((accumulator, funcFragment) => {
-        return {
-          ...accumulator,
-          [funcFragment.name]: newContract[funcFragment.name]
-        };
-      }, {});
-      setContract(contractInterface);
+      ) as T;
```

Comment:
Why do we have to cast this? T is extending contract so shouldn't this be inferred correctly?

Reply:
It wasn't possible and throwing the TypeScript errors though

![image](https://user-images.githubusercontent.com/45052332/147828497-a134989c-d8f7-4a8e-9408-b6a4ef0a0324.png)

![image](https://user-images.githubusercontent.com/45052332/147828504-032338d5-3d5e-4790-a71d-847a251e9fa7.png)

And the `Contract` class from ethers does not have a generic type argument, so it was pretty much a work-around